staging/src/k8s.io/apiserver/pkg/server/options/etcd.go

```diff
@@ -213,12 +227,48 @@ func (s *EtcdOptions) Complete(storageObjectCountTracker flowcontrolrequest.Stor
 	}
 
 	if len(s.EncryptionProviderConfigFilepath) != 0 {
-		transformerOverrides, kmsPluginHealthzChecks, err := encryptionconfig.LoadEncryptionConfig(s.EncryptionProviderConfigFilepath, stopCh)
+		ctx, closeTransformers := wait.ContextForChannel(stopCh)
+
+		transformerOverrides, kmsPluginHealthzChecks, configContentHash, kmsMaxTimeout, totalKMSPlugins, err := encryptionconfig.LoadEncryptionConfig(s.EncryptionProviderConfigFilepath, s.EncryptionProviderConfigAutomaticReload, ctx.Done())
 		if err != nil {
+			// in case of error, we want to close partially initialized (if any) transformers
+			closeTransformers()
+			return err
+		}
+
+		s.dynamicTransformers = encryptionconfig.NewDynamicTransformers(transformerOverrides, kmsPluginHealthzChecks, closeTransformers, kmsMaxTimeout, totalKMSPlugins)
+
+		if !s.EncryptionProviderConfigAutomaticReload {
+			s.kmsPluginHealthzChecks = kmsPluginHealthzChecks
+		}
+
+		// add post start hook to start hot reload controller
+		// adding this hook here will ensure that it gets configured exactly once
+		err = addPostStartHook(
+			"start-kms-encryption-config-hot-reloader",
+			func(hookContext server.PostStartHookContext) error {
+				// get context
+				// we ignore the cancel func because this context should only be canceled when stopCh is closed
+				ctx, _ := wait.ContextForChannel(hookContext.StopCh)
```

Comment:
Since the hot reload controller is triggered from the postStartHook I have propagated that context to tie the controller lifetime to the hook. For others like LoadEncryptionConfig (which also creates new gRPC services), and initialization of the dynamicTransformers I have kept the server lifetime context.

Reply:
We simplify this by passing the api-server stopCh to all the hot reload code paths to make sure they all are tied to the api-server lifecycle.